src/mapping/query/function/sha2.rs

```diff
@@ -0,0 +1,156 @@
+use super::prelude::*;
+use digest::DynDigest;
+use sha2::{Digest, Sha224, Sha256, Sha384, Sha512, Sha512Trunc224, Sha512Trunc256};
+
+#[derive(Debug)]
+pub(in crate::mapping) struct Sha2Fn {
+    query: Box<dyn Function>,
+    variant: Option<Box<dyn Function>>,
+}
+
+impl Sha2Fn {
+    #[cfg(test)]
+    pub(in crate::mapping) fn new(query: Box<dyn Function>, variant: Option<&str>) -> Self {
+        let variant = variant.map(|v| Box::new(Literal::from(Value::from(v))) as _);
+
+        Self { query, variant }
+    }
+}
+
+impl Function for Sha2Fn {
+    fn execute(&self, ctx: &Event) -> Result<Value> {
+        let bytes = required!(ctx, self.query, Value::Bytes(v) => v);
+
+        let variant = optional!(ctx, self.variant, Value::Bytes(v) => v)
+            .map(|v| String::from_utf8_lossy(&v).into_owned());
+
+        let mut digest: Box<dyn DynDigest> = match variant.as_deref() {
+            Some("SHA-224") => Box::new(Sha224::new()),
+            Some("SHA-256") => Box::new(Sha256::new()),
+            Some("SHA-384") => Box::new(Sha384::new()),
+            Some("SHA-512") => Box::new(Sha512::new()),
+            Some("SHA-512/224") => Box::new(Sha512Trunc224::new()),
+            Some("SHA-512/256") | None => Box::new(Sha512Trunc256::new()),
+            Some(v) => return Err(format!("unknown SHA-2 algorithm variant: '{}'", v)),
+        };
```

Comment:
This is one of my pain points of Rust - having to `Box<dyn..>` when the match arms return different objects of the same trait. It just feels like we are losing a (most likely Infinitesimally small) amount of efficiency by doing this.

Do you think it might be better to extract: 

```
    digest.update(&bytes);
    let sha2 = hex::encode(digest.finalize());
```
into a separate function and have each match arm call that?

That might increase binary size, but reduce runtime..

I realise I am splitting hairs here, but I am curious!

Reply:
I agree with your assessment. I considered the generic approach (after I wrote this implementation), but didn't think the extra allocation was worth going back and redoing things.

I think it's fine to get this in. If we get any proof in the future that there's a bottleneck here, we can revisit.

Reply:
I agree, it's highly unlikely to be a bottleneck. It's just one of those things I come up against a lot and keep wondering if there is a better way!

Reply:
I too dislike the `Box<dyn …>`, but a quick time test indicates the difference is lost in the noise, considering the much larger cost of the hash itself and the match on a string.

Reply:
I went ahead and removed the allocations 👍